services/users.js

```diff
@@ -82,8 +83,246 @@ module.exports = class UsersService {
+            assigned_by: assignedBy,
+            message,
+            created_at: Date.now(),
+          },
+        },
+      },
+      {
+        new: true,
+        runValidators: true,
+      }
+    );
+    if (user === null) {
+      user = await UserModel.findOne({id});
+      if (user === null) {
+        throw errors.ErrNotFound;
+      }
+
+      if (user.status.banned.status === status) {
+        return user;
+      }
+
+      throw new Error('ban status change edit failed for an unknown reason');
+    }
+
+    // Emit that the user ban status was changed.
+    await events.emitAsync(USERS_BAN_CHANGE, user, {
+      status,
+      assignedBy,
+      message,
+    });
+
+    return user;
+  }
+
+  static async setUsernameStatus(id, status, assignedBy = null) {
+    let user = await UserModel.findOneAndUpdate(
+      {
+        id,
+        status: {
+          $ne: status,
+        },
+      },
+      {
+        $set: {
+          'status.username.status': status,
+        },
+        $push: {
+          'status.username.history': {
+            status,
+            assigned_by: assignedBy,
+            created_at: Date.now(),
+          },
+        },
+      },
+      {
+        new: true,
+      }
+    );
+    if (user === null) {
+      user = await UserModel.findOne({id});
+      if (user === null) {
+        throw errors.ErrNotFound;
+      }
+
+      if (user.status.username.status === status) {
+        return user;
+      }
+
+      throw new Error(
+        'username status change edit failed for an unknown reason'
+      );
+    }
+
+    // Emit that the user username status was changed.
+    await events.emitAsync(USERS_USERNAME_STATUS_CHANGE, user, {
+      status,
+      assignedBy,
+    });
+
+    return user;
+  }
+
+  static async _setUsername(
+    id,
+    username,
+    fromStatus,
+    toStatus,
+    assignedBy,
+    resetAllowed = false
+  ) {
+    try {
+      const query = {
+        id,
+        'status.username.status': fromStatus,
+      };
+      if (!resetAllowed) {
+        query.username = {$ne: username};
+      }
+
+      let user = await UserModel.findOneAndUpdate(
+        query,
+        {
+          $set: {
+            username,
+            lowercaseUsername: username.toLowerCase(),
+            'status.username.status': toStatus,
+          },
+          $push: {
+            'status.username.history': {
+              status: toStatus,
+              assigned_by: assignedBy,
+              created_at: Date.now(),
+            },
+          },
+        },
+        {
+          new: true,
+        }
+      );
+      if (!user) {
+        user = await UsersService.findById(id);
+        if (user === null) {
+          throw errors.ErrNotFound;
+        }
+
+        if (user.status.username.status !== fromStatus) {
+          throw errors.ErrPermissionUpdateUsername;
+        }
+
+        if (!resetAllowed && user.username === username) {
+          throw errors.ErrUsernameTaken;
```

Comment:
Did you kick out `ErrSameUsernameProvided` and use `ErrUsernameTaken` on purpose?

Reply:
@wyattjoh

Reply:
Yes. We used it inconsistently before, so this reduces the amount of variation that you get from it.

Reply:
It just does not precisely describe the error situation here as it only occurs when I'm changing to my own username.